doc/design/design_high-available_sqlflow_server.md

```diff
@@ -0,0 +1,232 @@
+# High-available SQLFlow server
+
+## Motivations
+
+In the current system, SQLFlow client connects the SQLFlow server with a long live connection. the SQLFlow client sends
+a gRPC request which contains a SQL statement and blocking until the SQLFlow server finishes executing the SQL statement.
+
+For each SQL statement, the SQLFlow code generator would generate a submitter program in Python, and then the SQLFlow server
+would launch a process on the host or launch a distributed job on a cluster(Kubernetes/Yarn), this would cause two problems in the production environment:
+
+1. The local job may cause the SQLFlow server resource insufficient when there are too many SQL jobs.
+1. Sometimes, the SQL job takes too much time, and the gRPC calls timeout.
+1. If an SQLFlow server instance fails, the SQL jobs in this instance failed.
+
+In this design, we propose to:
+
+1. communicate with the SQLFlow server in a pooling manner.
+1. implement `ClusterJobRunner` on the server-side to launch the SQL Job on Kubernetes.
```

Comment:
implement -> Implement

Reply:
Why not directly use ElasitcDL?

Reply:
I prefer to call EDL API in the Kubernetes Pod because:
1.  SQLFlow server doesn't need to consider about the dependency for different submitter program, for example: the server can launch k8s Pod with `sqlflow:edl` Docker image to run EDL submitter program and `sqlflow:alps` Docker image to run ALPS submitter program.
1. We can deploy **ONE** SQLFlow server cluster and serve all the submitter type, just launch the SQLFlow Docker image with different submitter dependency installed.

Reply:
That's a bit complex, like we are going to do:

SQLFlow server -> call k8s API to start a submitter pod -> submitter call k8s api to start EDL master -> EDL master call k8s API to start workers -> submitter pod wait the EDL master to return -> SQLFlow server return.

Reply:
Yes, you're right, the k8s pod cmd can be:

``` bash
sqlflowcmd -e "sql...."
```